# Text Book: 
#An Introduction to Statistical Learning 
Gareth James • Daniela Witten • Trevor Hastie Robert Tibshirani

# Review and Questions from Class1


#**Two Purpose of the class:**
1. What is machine learning?
2. How to solve real life problems with machine learning?

## 1. **Introduction to Machine learning** 


*   Machine learning Applications:

1.   Submitting a query to a search engine,
2.   Imagine recognation.
3.   Recomender system.
4.   Automatic translation.
5.   Automatic scoring.
6.   Stock prediction.
*  Input---->Machine Learning----->output

### 1.1 **Input/Data**

Test responses, house price, stock prices each day, imagines, essays, a passage...

### 1.2 **Machine Learning Algorithms**

Define **a loss function** and find a way to minimize it with evaluation criteria.

Some examples of functions to be maximining or minimizing and algorithms:

1.Maximum likelihood estimation (MLE) is a method that determines values for the parameters of a model. The parameter values are found such that they maximise the likelihood that the process described by the model produced the data that were actually observed. See below.

2.Least squares estimaiton:
Least squares minimisation is another common method for estimating parameter values for a model in machine learning. It turns out that when the model is assumed to be Gaussian, the MLE estimates are equivalent to the least squares method.

3.Expectation-Maximization algorithms, an iterative solution: derivative for the likelihood function is hard to find, using EM is the way to go to maximizing the function:

E-Step. Estimate the missing variables in the dataset.
M-Step. Maximize the parameters of the model in the presence of the data.


4.Bayes theorem:

$P(A \mid  B) = P(B \mid A) * P(A)$.

We want to find $\theta$, so as to maximizing $P(X \mid \theta)$. It is the same as maximizing $P(\theta \mid  X)$, because 
$P(\theta \mid  X) = P(X \mid \theta) * P(\theta)$. Here $P(\theta)$ is a prior distribution for the parameter $\theta$. 

5.MCMC: Markov chain Monte Carlo (MCMC) methods comprise a class of algorithms for sampling from a probability distribution/ likelihood function: aiming for the sample matches the actual desired distribution.




###1.3 **Supervised vs. Unsupervised**

Supervised statistical learning: to build a statistical model 
for predicting, or estimating, with an output based on one or more inputs. Problems of this nature occur in fields as diverse as business, medicine, astrophysics, and public policy.  With independent, observable predictor variables $X$, there are responses Y, we need to fit a model that relates the response Y to the predictors X, with the aim of accurately predicting the response for future observations (called prediction) or better understanding the relationship between the response and the predictors (called inference).


Unsupervised statistical learning: there are inputs $X$ but no supervising output $Y$. We only have observale varibales $X$, and with the data, we are to learn relationships and structures between the variables $X$. Example: cluster users/customers.

For example, in a market segmentation study we might observe multiple characteristics (variables) for potential customers, such as zip code, family income, and shopping habits. We might believe that the customers fall into different groups, such as big spenders versus low spenders. If the information about each customer’s spending patterns were available, then a supervised analysis would be possible. However, this information is not available—that is, we do not know whether each potential customer is a big spender or not. In this setting, we can try to cluster the customers on the basis of the variables measured, in order to identify distinct groups of potential customers. Identifying such groups can be of interest because it might be that the groups differ with respect to some property of interest, such as spending habits.

Semi-supervised learning problem: there are some m observations with response Y, but there are n-m observations that have no response value; mnay times the collection of Y is expensive or difficult.

We will have applications for both supervised and unsupervised.

### 1.4 Output data
**bold text**
 Classification(image recognitation, disease, pass/fail, YES/No), Prediction(stock price, students attribute estimates, sale price)

## 2.**Statistics**

### 2.1 **Dependent Variable and Independent Variables.**

Let Y, a dependent variable, be an observed quantitative variable. $X = (X_{1},X_{2},...,X_{p})$ where $X_{1}, X_{2}, . . . , X_{p}$ are p independent variables. We assume that there is some relationship between Y and $X = (X_{1},X_{2},...,X_{p})$, which can be written in the very general form
$Y =f(X)+ϵ$. Here f is some fixed but unknown function of $X_{1}, X_{2}, . . . , X_{p}$, and $ϵ$ is a random error term, which is independent of X and has mean zero. In this formulation, f represents the systematic information that X provides about Y.

We may wish to estimate f: prediction and inference. 

For example, X=(maybe years of education, age), Y is his/her salary. Using observed data, we can use statistical tool to predict the salary Y for some people given their education and age. We may also wants to know how the years of education affect one's salary.

What is exactly $f(x)$? 

In many situations, we do not care (it is like a black box); we just need to have a good prediction/estimates. In other situations, we may want to know exactly what is $f(x)$, so that we know how those varibales $X = (X_{1},X_{2},...,X_{p})$ change affect our resppnses/outcome or the values of Y. What is the relationship between the response and each predictor?



How to find a good prediction for Y:
1. Aiming for the estimation of function $f(x)$. Try to find better model and methods, to reduce the estimation error.
2. However, there is some irreducible errors from $ϵ$. 

Therefore, our prediction would always have some error in it.





###2.2 **Statistical Models and Methods**

How to estimates $f(x)$
1. **Parametric Methods.** For example, we may assume
$f(X)=\beta_{0}+\beta_{1}*X_{1}+\beta_{2}*X_{2}+....\beta_{p}*X_{p}$. We need to estimate the unknown parameters $\beta_{0},....\beta_{p}$.

2. **Non-parametric Methods.** Here, we do not make explicit assumptions about the functional form of $f(x)$. We search for an estimate of $f(x)$ that gets as close to the data points as possible. We also want the estimates to be as smooth as possible.

Parametric Methods are quite easy to understand the relationship between Y and $X = (X_{1},X_{2},...,X_{p})$.
Non-parametric Methods are like black box, not easy to interprete.

We try to seek models and methods with the trade-off between flexibility and interpretability.

A simple, easy to understand and interpretable, less flexible model may have bigger errors, while a flexible, less to understand model may have smaller errors.

Two concepts we need to know in statistics: Mean and Variance of a random variable.

We define the mean of a random variable X as
$EX=∫xdp(x)$, and for a function $f$,
$E[f(X)]=∫ f(x)dp(x)$. X is a discrete random variable, then the mean or expection is $EX=∑_{x} xp(x)$.

The variance of a random variable
X is defined as $Var[X]=E(X-EX)^{2}$.

###2.3 **Bias and Variances**

How to measure the quality of the prediction?

In the regression setting, the
most commonly-used measure is the mean squared error (MSE), given by mean
\begin{align}
MSE = 1/n ∑_{i=1}^{n}(y_{i} −\hat{f}(x_{i}))^{2} ,
\end{align}

Many statistical methods specifically estimate coefficients so as to minimize the training set MSE. For these methods, the training set MSE can be quite small, but the test MSE is often much larger.

Training set: data that has both $X$ and $Y$, and are used to derive predictions or inferences. 

Test set: the predictions are derived based on the observable data $X$ uisng the model derived from the traing set data.
A good model is one that yield small test MSE.

In practice, cross-validation is used to select a method that minimizes the MSE for the validation data; hope the derived model will yield small test MSE.

For classification setting, the error rate is defined as
\begin{align}
MSE = 1/n ∑_{i=1}^{n}I(y_{i} != \hat{y}_{i})
\end{align}
where $I(y_{i} != \hat{y}_{i})$ is an indicator variable that equals 1 if $y_{i} != \hat{y}_{i}$   and zero otherwise.

A good classifier is one for which the test error above is smallest.

The expected test MSE has three parts: Bias ($\hat{f}(x_{i})$, Variance, and irreducible error term $var(ϵ)$. 
\begin{align}
MSE(\hat{\theta})=E(\hat{\theta}-\theta)^{2}=E(\hat{\theta}-E(\hat{\theta})+E(\hat{\theta})-\theta)^{2}
\end{align}
\begin{align}
=E(\hat{\theta}-E(\hat{\theta}))^{2}+E[E(\hat{\theta})-\theta]^{2}
\end{align}
\begin{align}
=VAR(\hat{\theta})+[E(\hat{\theta})-\theta]^{2}=Variance+SQUAREDBIAS
\end{align}


**The Bias-Variance Trade-Off**

Variance refers to the amount by which $\hat{f}$ would change if we estimated it using a different training data set. In general, more flexible statistical methods have higher variance. 

Bias refers to the error that is introduced by approximating a real life problem, which may be extremely complicated, by a much simpler model. For example, linear regression assumes that there is a linear relationship between Y and $X_{1}, X_{2}, . . . , X_{p}$. It is unlikely that any real life problem truly has such a simple linear relationship, and so performing linear regression will undoubtedly result in some bias in the estimate of f. 

As a general rule, as we use more flexible methods, the variance will increase and the bias will decrease.

In general, when inference is the goal, there are clear advantages to using simple and relatively inflexible statistical learning methods. However, we will often obtain more accurate predictions using a less flexible method too!
        
    

###2.5 **Regression Versus Classification Problems**

Variables can be characterized as either quantitative or qualitative (also known as categorical). 

Quantitative variables take on numerical values. Examples include a person’s age, height, or income, the value of a house, and the price of a stock. 

In contrast, qualitative variables take on values in one of K different classes, or categories. Examples of qualitative variables include a person’s gender (male or female), the brand of product purchased (brand A, B, or C), stock increase or decrease for centain day.
a person defaults on a debt (yes or no),
cancer diagnosis (Acute Myelogenous Leukemia, Acute Lymphoblastic Leukemia, or No Leukemia). 


We tend to refer to problems with a quantitative response as regression problems, while those involving a qualitative response are often referred to as classification problems. 


However, the distinction is not always that crisp. Least squares linear regression is used with a quantitative response, whereas logistic regression is typically used with a qualitative (two-class, or binary) response. As such it is often used as a classification method. But since it estimates class probabilities, it can be thought of as a regression.

We will have applications for both regression and classification.

## **First simple application of model fit and estimation, and likelihood function.**

Gaussian distribution has 2 parameters. The mean, μ, and the standard deviation, σ. Maximum likelihood estimation is a method that will find the values of μ and σ that result in the curve that best fits the data.
For observations $X_{1}, \cdots, X_{n}$ from Gaussian with mean, μ, and the standard deviation, σ, the probability of the event happened is

$L(X, \mu, \sigma)=P(X_{1},\cdots,X_{n} \mid \mu, \sigma)=∏_{i=1}^{n}P(X_{i} \mid \mu, \sigma)$.

To solove for the maximum, take log and take derivative swith respect to $\mu$ and $\sigma$, we will have the estimates
$\hat{\mu}=\overline{X}=\frac{1}{n}∑_{i=1}^{n}X_{i}$,
$\hat{\sigma}^{2}=Var(X)=\sum_{i=1}^{n} \frac{(X_{i}-\overline{X})^{2}}{n}$

$Var(\hat{\mu})=SE(\hat{\mu})=Var(\frac{1}{n}∑_{i=1}^{n}X_{i})=\frac{\sigma^{2}}{n}$

Normal density function is
$p(x)=\frac{1}{\sigma \sqrt{2\pi}}exp^{-\frac{1}{2}(\frac{x-\mu}{\sigma})^{2}}$

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns; sns.set()
import numpy as np
# For dataframes
import pandas as pd 


### Given a function, with known parameters, we can plot them: random draw samples from the distribution and plot the samples.
### How do we find the unknown parameters for a function, given the obervations of sample data from the function?

In [ ]:
x

In [ ]:

mu, sigma = 0, 10 # mean and standard deviation
n=1000
x = np.random.normal(mu, sigma, n)

count, bins, ignored = plt.hist(x, 30, density=True)
plt.plot(bins, 1/(sigma * np.sqrt(2 * np.pi)) *
               np.exp( - (bins - mu)**2 / (2 * sigma**2) ),
         linewidth=2, color='r')

plt.ylabel('density')
plt.xlabel('x')
plt.title("Normal Distribution")
plt.show()

In [ ]:
x.mean(),np.mean(x), np.std(x),x.max(), x.argmax()# returns indices of the max element of the array,



(0.00403202863844546,
 0.00403202863844546,
 1.0368005811497654,
 3.1922110414267384,
 55)

In [ ]:
X = np.random.uniform(-2,2, n)
X

In [ ]:
X = np.random.uniform(-2,2, n)
Z=2+3*X

In [ ]:
mu, sigma = 0, 2 # mean and standard deviation
n=100
𝜖 = np.random.normal(mu, sigma, n)
X = np.random.uniform(-2,2, n)
Y=2+3*X+𝜖
Z=2+3*X
Y[0]=9
df = pd.DataFrame({'x':X, 'y':Y,'z':Z})
plt.plot('x', 'z',data=df,color='red')
plt.scatter('x','y',data=df,color='blue')
plt.title("Linear Regression")
plt.show()

##**Overfitting and underfitting**

##np.random.randn(d0,d1),  Return a sample (or samples) from the “standard normal” distribution.
##np.random.rand(d0,d1),  Return a sample (or samples) of random values in (0,1) 

In [ ]:


def make_data(N=30, err=0.8, rseed=1):
    # randomly sample the data
    rng = np.random.RandomState(rseed)
    X = rng.rand(N, 1) ** 2
    y = 10 - 1. / (X.ravel() + 0.1)
    if err > 0:
        y += err * rng.randn(N)
    return X, y

In [ ]:
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.pipeline import make_pipeline

def PolynomialRegression(degree=2, **kwargs):
    return make_pipeline(PolynomialFeatures(degree),
                         LinearRegression(**kwargs))

In [ ]:
X, y = make_data()
#xfit = np.linspace(-0.1, 1.0, 1000)[:, None]
xfit=np.linspace(-0.1, 1.0, 1000).reshape(-1, 1)
model1 = PolynomialRegression(1).fit(X, y)
model20 = PolynomialRegression(20).fit(X, y)

fig, ax = plt.subplots(1, 2, figsize=(10, 6))
#fig.subplots_adjust(left=0.0625, right=0.95, wspace=0.1)

ax[0].scatter(X.ravel(), y, s=40)
ax[0].plot(xfit.ravel(), model1.predict(xfit), color='gray')
ax[0].axis([-0.1, 1.0, -2, 14])
ax[0].set_title('High-bias model: Underfits the data', size=14)

ax[1].scatter(X.ravel(), y, s=40)
ax[1].plot(xfit.ravel(), model20.predict(xfit), color='gray')
ax[1].axis([-0.1, 1.0, -2, 14])
ax[1].set_title('High-variance model: Overfits the data', size=14)
plt.show()

## 3. Computer programing

3.1 Python, R, Java, C, C++

3.2 Define a function that need to be optimized: minimum or maximinum

3.3 Algorithem to solve the optimized problem.

3.4 Measurement error

##**HW Project, Practice python**

###HW, read Introduction to Styatistical learning, chapter 1

1. For each of parts (a) through (d), indicate whether we would generally expect the performance of a flexible statistical learning method to be better or worse than an inflexible method. Justify your answer.

(a) The sample size n is extremely large, and the number of predictors p is small.

(b) The number of predictors p is extremely large, and the number of observations n is small.

(c) The relationship between the predictors and response is highly non-linear.

(d) The variance of the error terms, i.e. σ2 = Var(ε), is extremely high.


2. Explain whether each scenario is a classification or regression problem, and indicate whether we are most interested in inference or prediction. Finally, provide n and p.

(a) We collect a set of data on the top 500 firms in the US. For each firm we record profit, number of employees, industry and the CEO salary. We are interested in understanding which factors affect CEO salary.

(b) We are considering launching a new product and wish to know whether it will be a success or a failure. We collect data on 20 similar products that were previously launched. For each product we have recorded whether it was a success or failure, price charged for the product, marketing budget, competition price, and ten other variables.

(c) We are interesting in predicting the % change in the US dollar in relation to the weekly changes in the world stock markets. Hence we collect weekly data for all of 2012. For each week we record the % change in the dollar, the % change in the US market, the % change in the British market, and the % change in the German market.

3. Describe the differences between a parametric and a non-parametric statistical learning approach. What are the advantages of a parametric approach to regression or classification (as opposed to a nonparametric approach) ? What are its disadvantages ?

4. What are the advantages and disadvantages of a very flexible (versus a less flexible) approach for regression or classification ? Under what circumstances might a more flexible approach be preferred to a less flexible approach? When might a less flexible approach be preferred ?


### Answer

1.

a)A flexible method will fit the data closer and with the large sample size, would perform better than an inflexible approach.

b)A flexible method would overfit the small number of observations, so worse then inflexible method.

c) With more degrees of freedom, a flexible method would fit better than an inflexible one.

d) A flexible method would fit to the noise in the error terms and increase variance.


2.
a)Regression and inference with n=500 and p=3

b) Classification and prediction with n=20 and p=13

c) Regression and prediction with n=52 and p=3

3.
A parametric approach reduces the problem of estimating f down to one of estimating a set of parameters because it assumes a form for f.

A non-parametric approach does not assume a patricular form of f and so requires a very large sample to accurately estimate f.

The advantages of a parametric approach to regression or classification are the simplifying of modeling f to a few parameters and not as many observations are required compared to a non-parametric approach.

The disadvantages of a parametric approach to regression or classification are a potentially inaccurate estimate f if the form of f assumed is wrong or to overfit the observations if more flexible models are used.

4.

The advantages of a very flexible approach are that it may give a better fit for non-linear models and it decreases the bias.

The disadvantages of a very flexible approach are that it requires estimating a greater number of parameters, it follows the noise too closely (overfit) and it increases the variance.

A more flexible approach would be preferred to a less flexible approach when we are interested in prediction and not the interpretability of the results.

A less flexible approach would be preferred to a more flexible approach when we are interested in inference and the interpretability of the results.
